Create pytestBDD_reporter.py

bbornhau opened the pull request:
decorator to enable gherkin style reporting when using pytest_bdd
To enable the following functions have to be annotated in conftest.py:
```python
@pytestBDD_reporter()
def pytest_bdd_after_step(request, feature, scenario, step, step_func,step_func_args):
	pass

@pytestBDD_reporter()
def pytest_bdd_after_scenario(request, feature, scenario):
	pass

@pytestBDD_reporter()
def pytest_bdd_step_error(request, feature, scenario, step, step_func, step_func_args, exception):
	pass

@pytestBDD_reporter()
def pytest_bdd_before_scenario(request, feature, scenario):
	pass
```

bbornhau commented:
@vitalybu Looks like the build is blocked due an error caused but the agent.
Is there anything i can/have to do to solve this?

vitalybu commented:
> @vitalybu Looks like the build is blocked due an error caused by the agent.
> Is there anything i can/have to do to solve this?

Nothing you can do at this point. The CI needs to be updated to handle external PRs.
Will be fixed in a few days.

bbornhau commented:
> > @vitalybu Looks like the build is blocked due an error caused by the agent.
> > Is there anything i can/have to do to solve this?
> 
> Nothing you can do at this point. The CI needs to be updated to handle external PRs.
> Will be fixed in a few days.

Hi there, looks like it is still broken. Any ETA in this? I would like to start a blog post about the integration with pytest-bdd but I would need to wait until this has been merged...

vitalybu commented:
> > > @vitalybu Looks like the build is blocked due an error caused by the agent.
> > > Is there anything i can/have to do to solve this?
> > 
> > 
> > Nothing you can do at this point. The CI needs to be updated to handle external PRs.
> > Will be fixed in a few days.
> 
> Hi there, looks like it is still broken. Any ETA in this? I would like to start a blog post about the integration with pytest-bdd but I would need to wait until this has been merged...

Hi @bbornhau, it indeed takes a bit more time than originally anticipated.
I hope everything will be resolved within a couple of days.

vitalybu commented on src/testproject/decorator/pytestBDD_reporter.py:
```diff
@@ -0,0 +1,96 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from src.testproject.helpers.activesessionhelper import (get_active_driver_instance)
+import os
+from decorator import decorator
```
I would reorder the imports to have the `import` statements at the top and the `from` below:
```python
import os

from decorator import decorator
from src.testproject.helpers.activesessionhelper import (get_active_driver_instance)
```

vitalybu commented on src/testproject/decorator/pytestBDD_reporter.py:
```diff
@@ -0,0 +1,96 @@
+    otherwise
+    (False) only on failure.
+    """
+    # set environemnt bariable TP_ENABLED to False to disable reporting into
+    # test
+    # project
+    if os.getenv("TP_ENABLED") == "False":
+        return pytestBDD_reporter
```
Not sure what is the purpose of this variable.
If one doesn't want to report to TP, the decorator can be removed from the tests, no?
We don't have this setting in any other "plugin".

bbornhau replied:
Hi @vitalybu 
thanks for the comments.
The idea behind this is to easily deactivate the reporting into Testproject while developing the tests.
Removing the decorator is tedious and error prone so i thought it would be a good idea to have this set like this.
I added a command line parameter to quickly set this in my tests.

mstrelex replied:
@vitalybu 
I'm fine with it. But we will have to align then all SDKs to this behavior.

vitalybu commented on src/testproject/decorator/pytestBDD_reporter.py:
```diff
@@ -0,0 +1,96 @@
+    if os.getenv("TP_ENABLED") == "False":
+        return pytestBDD_reporter
+
+    if os.getenv("TP_DISABLE_AUTO_REPORTING") != "True":
+        os.environ["TP_DISABLE_AUTO_REPORTING"] = "True"
```
This if statement is redundant. You can always just set the variable to `True`:
`os.environ["TP_DISABLE_AUTO_REPORTING"] = "True"`

bbornhau replied:
Hi @vitalybu i copied this one from the behave_bdd source code....and yes, if i look at it that way you are right.
Will do the changes

vitalybu replied:
Thanks. @Rantzur1992 please implement these changes in the Behave code as well.

vitalybu commented on src/testproject/decorator/pytestBDD_reporter.py:
```diff
@@ -0,0 +1,96 @@
+    # for step error:
+    #      (request, feature, scenario, step, step_func,step_func_args,
+    #       exception)
+    if hook_name == "pytest_bdd_after_step":
+        driver = get_active_driver_instance()
+        step = args[3]
+        message = "{} {}".format(step.keyword, step.name)
+        report_step(driver=driver,
+                    step=step,
+                    screenshot=screenshot,
+                    message=message
+                    )
+
+    if hook_name == "pytest_bdd_step_error":
+        driver = get_active_driver_instance()
+        step = args[3]
+        exception = args[6]
+        step.failed = True
+        message = "{} {}".format(type(exception).__name__, str(exception))
+        report_step(driver=driver,
+                    step=step,
+                    screenshot="True",
+                    message=message
+                    )
+    if hook_name == "pytest_bdd_after_scenario":
+        driver = get_active_driver_instance()
+        scenario = args[2]
+        report_test(driver=driver,
+                    scenario=scenario
+                    )
```
Assuming only one hook can be processed at a time, I think that this code can be restructured as:
```python
if hook_name == "pytest_bdd_after_step":
    ...
elif hook_name == "pytest_bdd_step_error":
    ...
elif hook_name == "pytest_bdd_after_scenario":
    ...
```

bbornhau replied:
Hi, @vitalybu absolutely right.
By the way, this again is the same structure as i found in the behave_bdd plugin, hence i propose to change it there also.

vitalybu replied:
@Rantzur1992, this change as well.

Rantzur1992 replied:
Done https://github.com/testproject-io/python-opensdk/pull/154

vitalybu reviewed (commented):
While we're working on fixing the CI, I reviewed your PR and added some comments

bbornhau commented:
@vitalybu  looks like the CI does not kick off...

vitalybu commented:
@bbornhau, I have just introduced another update to the CI that should skip running the problematic workflows on external (fork) pull requests.
I suggest squashing the commits on your branch and rebasing to the latest master.

bbornhau commented:
> @bbornhau, I have just introduced another update to the CI that should skip running the problematic workflows on external (fork) pull requests.
> I suggest squashing the commits on your branch and rebasing to the latest master.

@vitalybu as you might see, i am trying sign-off my commit, same way as i did before, but it is not registering it...

vitalybu commented:
> > @bbornhau, I have just introduced another update to the CI that should skip running the problematic workflows on external (fork) pull requests.
> > I suggest squashing the commits on your branch and rebasing to the latest master.
> 
> @vitalybu as you might see, i am trying sign-off my commit, same way as i did before, but it is not registering it...

You're trying to sign the https://github.com/testproject-io/python-opensdk/pull/141/commits/322cfde7150849e86aa9bd9d71f9211935660ee4 commit?
That's the one missing the signature based on the DCO job.

If you'll squash all your commits, you'll have only one to deal with :)

bbornhau commented:
> > > @bbornhau, I have just introduced another update to the CI that should skip running the problematic workflows on external (fork) pull requests.
> > > I suggest squashing the commits on your branch and rebasing to the latest master.
> > 
> > 
> > @vitalybu as you might see, i am trying sign-off my commit, same way as i did before, but it is not registering it...
> 
> You're trying to sign the [322cfde](https://github.com/testproject-io/python-opensdk/commit/322cfde7150849e86aa9bd9d71f9211935660ee4) commit?
> That's the one missing the signature based on the DCO job.
> 
> If you'll squash all your commits, you'll have only one to deal with :)

Hi @vitalybu , 
by the way, thanks for your patience, as you can see i am clearly not a GIT master :)
So i squashed and pushed the changes up. Now the format test fails...but the DCO is ok ...
I guess i'll (re)tackle the 'black .' check next :)

vitalybu commented:
> > > > @bbornhau, I have just introduced another update to the CI that should skip running the problematic workflows on external (fork) pull requests.
> > > > I suggest squashing the commits on your branch and rebasing to the latest master.
> > > 
> > > 
> > > @vitalybu as you might see, i am trying sign-off my commit, same way as i did before, but it is not registering it...
> > 
> > 
> > You're trying to sign the [322cfde](https://github.com/testproject-io/python-opensdk/commit/322cfde7150849e86aa9bd9d71f9211935660ee4) commit?
> > That's the one missing the signature based on the DCO job.
> > If you'll squash all your commits, you'll have only one to deal with :)
> 
> Hi @vitalybu ,
> by the way, thanks for your patience, as you can see i am clearly not a GIT master :)
> So i squashed and pushed the changes up. Now the format test fails...but the DCO is ok ...
> I guess i'll (re)tackle the 'black .' check next :)

I see that your commit now includes changes that were not originally part of your branch.
Maybe at this point a better approach would be to create a new branch, commit only your changes, make sure it passes flake8 and black locally and then open a new PR.

bbornhau commented:
> > > > > @bbornhau, I have just introduced another update to the CI that should skip running the problematic workflows on external (fork) pull requests.
> > > > > I suggest squashing the commits on your branch and rebasing to the latest master.
> > > > 
> > > > 
> > > > @vitalybu as you might see, i am trying sign-off my commit, same way as i did before, but it is not registering it...
> > > 
> > > 
> > > You're trying to sign the [322cfde](https://github.com/testproject-io/python-opensdk/commit/322cfde7150849e86aa9bd9d71f9211935660ee4) commit?
> > > That's the one missing the signature based on the DCO job.
> > > If you'll squash all your commits, you'll have only one to deal with :)
> > 
> > 
> > Hi @vitalybu ,
> > by the way, thanks for your patience, as you can see i am clearly not a GIT master :)
> > So i squashed and pushed the changes up. Now the format test fails...but the DCO is ok ...
> > I guess i'll (re)tackle the 'black .' check next :)
> 
> I see that your commit now includes changes that were not originally part of your branch.
> Maybe at this point a better approach would be to create a new branch, commit only your changes, make sure it passes flake8 and black locally and then open a new PR.

Hi, yeah had the same idea when i saw the log entries. Although all passed now :) 
The black issue was on one of my files, but yeah, let's start fresh :).
So i'll delete the branch ( i guess )  and will create a new one.

vitalybu commented:
> > > > > > @bbornhau, I have just introduced another update to the CI that should skip running the problematic workflows on external (fork) pull requests.
> > > > > > I suggest squashing the commits on your branch and rebasing to the latest master.
> > > > > 
> > > > > 
> > > > > @vitalybu as you might see, i am trying sign-off my commit, same way as i did before, but it is not registering it...
> > > > 
> > > > 
> > > > You're trying to sign the [322cfde](https://github.com/testproject-io/python-opensdk/commit/322cfde7150849e86aa9bd9d71f9211935660ee4) commit?
> > > > That's the one missing the signature based on the DCO job.
> > > > If you'll squash all your commits, you'll have only one to deal with :)
> > > 
> > > 
> > > Hi @vitalybu ,
> > > by the way, thanks for your patience, as you can see i am clearly not a GIT master :)
> > > So i squashed and pushed the changes up. Now the format test fails...but the DCO is ok ...
> > > I guess i'll (re)tackle the 'black .' check next :)
> > 
> > 
> > I see that your commit now includes changes that were not originally part of your branch.
> > Maybe at this point a better approach would be to create a new branch, commit only your changes, make sure it passes flake8 and black locally and then open a new PR.
> 
> Hi, yeah had the same idea when i saw the log entries. Although all passed now :)
> The black issue was on one of my files, but yeah, let's start fresh :).
> So i'll delete the branch ( i guess ) and will create a new one.

Please do and create a new pull request.
I'll close this one.
